### ISSUE TITLE
Can't produce a good shape model

### ISSUE THREAD

**Issue by bonoshunki** [NVlabs/nvdiffrec#84]
Hi, recently I’ve been working on creating 3D models from Images using Nvdiffrec. I’m working on a banana model right now because I though simple objects are easy to duplicate.

But I can’t make a good banana model. Actually the model looks good in some directions, but the others are not. I changed some configs, but the result didn’t change significantly. Could you give me some advise?

Best regards,
----------------------------------------------------------------------------------------------------------------------------------
I’ll also post the situations.

- images
   - I took 189 banana photos. The half of them is from the side (around 360 degree), and the other half is from top. 
   - Like these.
<img width = "300px" src="https://user-images.githubusercontent.com/63286236/196862887-84318d68-5a33-47fd-af8c-d64d863015fc.jpg"><img width = "300px" src="https://user-images.githubusercontent.com/63286236/196862170-9ed1a47e-164e-410a-8b5c-d0b9fb20c53b.jpg">

- masks
  - I created masks by using adobe photoshop. As long as I see, the quality isn’t bad.
  - Like these.
<img width = "300px" src="https://user-images.githubusercontent.com/63286236/196862286-e13c168a-95c4-4c76-bf7a-3545f3ab2c2d.jpg"><img width = "300px" src="https://user-images.githubusercontent.com/63286236/196862865-ab83c2b0-1d66-424f-99bd-05ba8db951fb.jpg">

- method
  - I read #3 issue and I did as written there. 

- config
  - Basically I used options written in `nerd_ehead.json`, but I reduced batch from 8 to 4 due to a GPU Memory problem. 
  - Because I read #18, #37, #42 and others, I changed some configs and then tried, but the result didn’t seem to be getting better.
  - Like this.
```
"random_textures": true,
"iter": 5000,
"save_interval": 100,
"texture_res": [2048, 2048],
"train_res": [1024, 1024],
"batch": 4,
"learning_rate": [0.03, 0.01],
"kd_min": [0.03, 0.03, 0.03],
"kd_max": [0.8, 0.8, 0.8],
"ks_min": [0, 0.08, 0],
"ks_max": [0, 1.0, 1.0],
"dmtet_grid": 128,
"mesh_scale": 5,
"background": "black",
"display": [{"bsdf": "kd"}, {"bsdf": "ks"}, {"bsdf": "normal"}],
"lock_pos": false,
"laplace_scale": 3000,
"laplace": "relative",
```
- result
<img width="300" alt="スクリーンショット 2022-10-20 14 07 18" src="https://user-images.githubusercontent.com/63286236/196862049-a6f95f5f-818d-4513-9f2b-953fa8d6fe9b.png">

<img width="300" src="https://user-images.githubusercontent.com/63286236/196863374-3bb5d546-3f6b-4f1e-a020-96d98ff78016.png"><img width="300" src="https://user-images.githubusercontent.com/63286236/196863369-72696f2d-c37d-4510-8666-c772bd072cf6.png">

<img width="300" src="https://user-images.githubusercontent.com/63286236/196863379-3716d66e-b93d-4c6e-b9f9-e083dce6de76.png"><img width="300" src="https://user-images.githubusercontent.com/63286236/196863377-51315e6b-64cf-4521-a10e-4d666e441f61.png">

<img width="300" src="https://user-images.githubusercontent.com/63286236/196863384-c2ee3a7d-7c75-4ced-b52a-9e86aefc47f9.png"><img width="300" src="https://user-images.githubusercontent.com/63286236/196863382-73d0a923-611d-43b1-9124-9e21f52cfce4.png">

**Comment by jmunkberg** [comment]
Hello @bonoshunki 

Yes, that looks worse than expected. With 189 images with good 3D coverage, I would expect higher quality, in line with the other results in the paper, such as nerd_gold.json example, which is essentially the same setup. 

You may have already checked this, but just in case: 
1. Verify the accuracy of the camera poses. I assume you used colmap? We sometimes check the poses by creating a NeRF using https://github.com/NVlabs/instant-ngp as a sanity check. It is likely the culprit here. We assume perfect poses, and do not correct for inaccuracies, so if the poses are bad, nvdiffrec will fail.
2. Accuracy of masks. The examples above look reasonable, so likely not the issue. 
3. Is the lighting constant in all images, and the object is static?

**Comment by bonoshunki** [comment]
@jmunkberg 
Hi,  thank you for reply! I appriciate your help.

1. Yes, I used colmap. Actually I runned [colmap2poses.py](https://gist.github.com/Sazoji/e20835d652c51f305ce328342af7fefd/revisions) created by Sazoji. And I'll post results later, but the NeRF results seem not to be bad when I compared the photos I took and NeRF produced.
2. Actually some masks don’t look good. So I tried without bad masks, but the result was the same.
3. I guess yes. I put the banana on a small table and fixed a light on the banana, then I took photos by moving camera, not banana.

---

- bad mask examples
  <img width="300" src="https://user-images.githubusercontent.com/63286236/196945436-f1bd43c9-47f4-4df4-aaa6-dbc704497d64.jpg"><img width="300" src="https://user-images.githubusercontent.com/63286236/196945403-a7649534-da70-4f29-ac4b-a0d1b1883c8e.jpg">

- NeRF results
  <img height="400" src="https://user-images.githubusercontent.com/63286236/196945263-676100d8-044e-4074-888c-cc5a9b4e427c.jpg"><img height="400" src="https://user-images.githubusercontent.com/63286236/196945295-e350c2f1-578b-43cd-9bdb-03372ead26b2.jpg">
<img height="400" src="https://user-images.githubusercontent.com/63286236/196946383-3190f9b2-853e-423e-9ce2-e27748649d67.jpg"><img height="400" src="https://user-images.githubusercontent.com/63286236/196946524-1d88e1b7-973c-441e-b519-5c13cb68ebea.jpg">
- NeRF scripts
```
# these options from nvdiffrec
colmap feature_extractor --database_path "$DATASET_PATH/database.db" \
    --image_path "$DATASET_PATH/images" \
    --ImageReader.single_camera 1 \
    --ImageReader.mask_path "$DATASET_PATH/masks"

colmap exhaustive_matcher --database_path "$DATASET_PATH/database.db"

mkdir "$DATASET_PATH/sparse"

colmap mapper --database_path "$DATASET_PATH/database.db" \
    --image_path "$DATASET_PATH/images/" \
    --output_path "$DATASET_PATH/sparse/" \
    --Mapper.num_threads 16 \
    --Mapper.init_min_tri_angle 4 \
    --Mapper.multiple_models 0 \
    --Mapper.extract_colors 0

# Run NeRF
colmap bundle_adjuster --input_path "$DATASET_PATH/sparse/0" \
    --output_path "$DATASET_PATH/sparse/0" \
    --BundleAdjustment.refine_principal_point 1

mkdir "$DATASET_PATH/colmap_text"

colmap model_converter --input_path "$DATASET_PATH/sparse/0" \
    --output_path "$DATASET_PATH/colmap_text" \
    --output_type TXT

python scripts/colmap2nerf.py --colmap_matcher exhaustive \
    --images "$DATASET_PATH/images" \
    --colmap_db "$DATASET_PATH/colmap.db" \
    --text "$DATASET_PATH/colmap_text" \
    --out "$DATASET_PATH/transforms.json"

cd "$DATASET_PATH"

python "$SCRIPT_DIR/scripts/colmap2nerf.py" --colmap_matcher exhaustive \
    --images images \
    --colmap_db colmap.db \
    --text colmap_text \
    --out transforms.json

cd "$SCRIPT_DIR"

python scripts/run.py --scene="$DATASET_PATH/" \
    --mode=nerf \
    --screenshot_transforms="$DATASET_PATH/transforms.json" \
    --screenshot_w=1024 \
    --screenshot_h=1024 \
    --screenshot_dir="$DATASET_PATH/screenshots" \
    --save_snapshot="$DATASET_PATH/snapshot.msgpack" \
    --n_steps=1000
```

**Comment by YusukeTosa** [comment]
I am also having the same problem.
By the way, are the optimal parameters robust for the object?
We have verified that there is no problem with the mask image and camera poses, so I suppose the matter is the hyperparameter tuning.

**Comment by yezifeiafei** [comment]
I met the same problem.
@YusukeTosa Can you share the way you used to verify the camera poses?
I estimated my camera poses, and don't know they are right or not?

**Comment by IvanGarcia7** [comment]
I am also having the same problem :( . How can i correctly estimate the camera poses?

**Comment by IvanGarcia7** [comment]
@bonoshunki Could you share the dataset to perform some test / experiments??

**Comment by TikZSZ** [comment]
Any solutions yet?? I'm running into similar problem

**Comment by Rickyeeeeee** [comment]
I also encountered the same problem, but I was able to **solve it** at the end! In my case the reason is poses incorrectness.
